### ISSUE TITLE
Prevent how tile reading interrupts itself while running over multiple entities

### ISSUE THREAD

**Issue by LevFendi** [Factorio-Access/FactorioAccess#219]
This is most evident when in smooth walking mode you run along a belt or over an ore patch.

**Comment by LevFendi** [comment]
The solution for this is potentially easy: If it has been less than X ticks since the last time you called `read_tile` then mute this call but still count it as a call.

The value for x is determined by the base value for the running speed. Therefore holding down the running key will no longer cause this issue.

This mute feature is overridden every time you press a move key so that rapidly pressing move keys will still allow self-interrupting output.

**Comment by ahicks92** [comment]
I would prefer an announcement on an edge trigger. That is when going off a transport belt to dirt that should be read always.

I think the longer term solution is earcons, the name for short sounds representing UI actions (or game, in this case).  If we end run around speech and just have a "walking on x" set of sounds the entire problem goes away. Of course, getting a bunch of sounds and figuring out how to get that working with mods is a bigger problem than the short term solutions so lol.  But, in my opinion that's what we'd ideally do.

**Comment by LevFendi** [comment]
Okay this can be achieved using a check that uses `on_tick`:
- Every time printout is called, it checks if it has been 30 ticks since the last call. If yes then this call is skipped and it raises the flag print_skipped. Afterward it logs itself as the latest call.
- On every 30th tick the game checks if a print has been skipped more than 60 ticks ago. If yes, the skipped flag is lowered and the last print statement is released.

- Therefore any new print within 30 ticks is skipped until there is a 60 tick pause. When running across a belt, this results in the start and end of the belt to be announced.

**Comment by ahicks92** [comment]
That seems overly complex.  All we have to do is lift announcing to a module which we should do anyway and just if prev_terrain != terrain.  Also if you're being literal about printout please let's not put it there.  If you want a reusable edge trigger abstraction ask and give me 10minutes or something.  Weirdly I have done that very specific thing a few times for other projects.

To be honest I'm skeptical that announcing without changes is worth it.  Like you announce on change and not on change and now an announcement has two meanings.  What do we think of modifying character reading to be "character, on grass", instead of announcing every so often?  I don't think anyone will care about periodic reminders and if they do let's skip to earcons, I can mostly take that on, got that friend of mine who's willing to help do sound editing for credit.

Also most terrains are pretty unimportant too.  No need to care much about grass versus desert.

I suspect as you do your blind playthrough you'll come to understand why I keep coming back to using unique footstep sounds rather than speech (hint: basically the same reason as flying text with a side of please stop being verbose).

**Comment by LevFendi** [comment]
This bug has nothing to do with reading the terrain, but I see where you are going with this while expanding the scope here. Feel free to experiment and share.

**Comment by ahicks92** [comment]
Yeah it does.  KK gets stuck reading terrain all the time.  Or "blank blank blank".  Or "grass blank grass blank grass" or rarely just "grass grass grass"

Maybe the underlying cause is a tighter scope and actually specifically related to entities.  The UX isn't though.  I haven't dug into this but if it's simpler than I think wrt internals then let's not widen the scope here, that can be a different issue at a different time.
